### ISSUE TITLE
We need a library to support hacks of different knitting machines, that can be used by applications. How to implement the abstraction layer?

### ISSUE THREAD

**Issue by mariobehling** [fossasia/2018.fossasia.org#3]
One of our goals this summer - a library that supports different hacks and can be used by different projects (through an abstraction layer). How should that abstraction look like? What technologies? How can we implement it?

**Comment by shiluka** [comment]
Hi @mariobehling ,

>How should that abstraction look like?

I would like to suggest abstraction of library as separate layers

1. serial communication 
2. for machine types 
3. utility functions
4. image handling functions
5. file handling functions

**Comment by mariobehling** [comment]
Which projects (knittic, ayab, sconklin hack, pfaff hack) can be implemented in the library? What are the limits?

**Comment by mariobehling** [comment]
@shiluka What technologies do you propose?

**Comment by shiluka** [comment]
@mariobehling ,

As I mentioned earlier abstraction of the library can be divided into separate layers such as serial communication, machine types, image handling functions, file handling functions, utility functions and exception handling functions.

I suggest to write core of the library in c++ language because using c++ hardware can be directly called. c++ library can be bind to python using [ctypes.](https://docs.python.org/2/library/ctypes.html)

Suggestions on this approach would be greatly helpful.

Thank you.

**Comment by tian2992** [comment]
Hello @shiluka, as I mentioned on #4 I think we should leverage existing code, which is mostly written in Python. We also need a flexible enough language that allows for both Web API support, and for low level serial communications, and image manipulation, which are great areas for Python. 

I agree that we should separate more layers, in particular separating a generic machine controller that allows for separation between view (QT GUI or Web), and the actual hardware, as well as to enable varied processing features. As @sameeragunarathne comments on #4 we agree on the need for API endpoints that can be flexible enough to connect and to manipulate using a similar (or hopefully a modular unified) interface.  In particular as @chris007de and I mention on #6 we also need a generic container for holding Image data and Metadata for the knitting process and operation.

**Comment by shiluka** [comment]
Hi @tian2992 , thanks for the comment. As you said ayab and sconklin is written in Python. But knittic is written in java. That's why I thought to implement core of the library in c++ and wrap it. Can you provide more details regarding the use of library for web API and generic container ?

**Comment by chris007de** [comment]
Hi,
I agree with @tian2992 's proposal of leveraging/re-using existing code to save time. We do not have to re-invent the wheel, rather we will have to build on existing foundations to be able to keep up with the tight schedule of GSoC.

The architecture we thought about for the AYAB software is already pretty good, but perhaps we should focus on separation of the tasks done in such an application, just like @shiluka already described in his post.

From my understanding, we have the following sections:
## Pattern Generation/Modification = Pattern Editor
(done in Knitic, not yet in AYAB)
This may be done in a completely different piece of software, but may also be integrated into the knitting application (just another mode of operation).
But remember: Generating patterns can be very sophisticated (which may also yield in sophisticated software), so it is a good idea to keep this separated.
Most basic method is to click some pixels, like the solution in Knitic. Experienced operators want to copy/paste areas, insert templates, mirror, rotate, ... - and of course, want to generate the metadata which contains information about used carriages etc (can be knitting machine model dependent).
The output of this part is a pattern (file/container), as described in https://github.com/fashiontec/knitapps/issues/6

## Knitting
Keep this part pretty straight forward, as the operator has to concentrate on her/his work. Some possibilities for modification of the pattern are OK, but only the final tweaks.
Sophisticated editing has to happen in the pattern editor.
Then, we already get into the topic of different machines/hacks. Each machine type needs/allows different settings, so we have to present different, machine dependent settings to the user (just as done by the AYAB software).
For the knitting operation itself, of course also different protocols and state machines are applicable for different machines and hacks.
The information presented to the user may differ slightly, but is basically the same. Here, it is important to present the right info to the operator, and not just *everything* (it's the knitting operation, not a debug mode).

So, basically, I am pretty satisfied with the plugin-based design done in the AYAB software, and where the separation of different machine models takes part. A plugin contains the physical and protocol layer (state machines) of a machine/hack, but also the proprietary settings which are applicable to each machine/hack.
For my understanding, the targeted knitting machine library could instantly be built on the foundation of the already existing AYAB plugin system. It might need some extensions here and there, but basically, it describes everything we need.

**Comment by tian2992** [comment]
Knittic uses Processing, which is a simplified language, based on Java. I think it can be easier to implement just that on Python rather than re-build everything in C++.

**Comment by tian2992** [comment]
Thanks @chris007de I also agree with your points.
The current AYAB architecture can support basically any kind of plugin for knitting, the main architecture issue we face is that currently the plugins handle both the internal layers of the protocol stack and settings, and the visual interface. It works fine for local printing, but if we want to implement web interfaces, we need to separate in order to support regardless of the interface.

**Comment by shiluka** [comment]
Hi @chris007de , 
Thank you. As you mentioned, for the initial start we can build the library on the foundation of the already existing ayab plugin system. 

Though I thought to create the library in C++ as @tian2992 and @chris007de mentioned, with the tight schedule of the gsoc leveraging current code system would be a good idea, so I’m going to develop the library using python.

These are the initial suggestions for the Library

<h1> Library to support knitting machines
	
**Basic abstraction of the library**

As we all agree with this basic abstraction, I would like to create abstraction of the library as separate layers
* serial communication : open, configure, read and write to serial port
* machine definitions : machine initialization, load to machine, save from machine
* image handling functions : memory allocation, read image file, setter and getter for image pixels
* file handling functions : read / write access
* common utility functions

**Add dependencies for the library**

There are several dependencies which ayab has such as pillow, pyserial, wsgiref and fysom with relevant versions. These dependencies should be added to the library. 

**Functions for the library**

* For access hardware directly
       * Add functions into library which interface with hardware
       * Use libraries to deal with serial ports
       * Ability to get all the functions via library 

* Emulate file formats : implement functions to emulate file formats.

* Add Arduino support 


**Installation**

Installation should be well documented in step by step. Additionally I suggest to add GUI based installation with usability improvements and standard installation procedures. Also without installing packages separately user should be able to installed them inside the library installation. 

* Installation on Linux (32bit, 64bit)
       * Prerequisites
       * Setup
* Installation on Windows (32bit, 64bit)
       * Prerequisites
       * Setup


**Tests on the library** 

* We can use a python test framework like [unittest](https://docs.python.org/2/library/unittest.html) to test the python library. It supports test automation and aggregation of the tests into collection.

* Tests on whether the library build correctly by trying to import functions from the library. 

* Test the installation

**Documentation of the library** 

Documentation on Class definitions, function definitions and links to the given source codes (Library overview). 


@mariobehling , @tian2992 and @chris007de  , Thanks very much for the valuable suggestions you have given.
